### PR TITLE
[Enhancement]: dynamodb: add arg to set deletion protection for table replica

### LINKED ISSUE #30213: [Enhancement]: Enable Deletion Protection for DynamoDB Table Replicas
### Description

Please add the ability to enable deletion protection for Dynamo DB Table Replicas, to match the functionality for tables themselves that was added in #29924.

### Affected Resource(s) and/or Data Source(s)

aws_dynamodb_table_replica

### Potential Terraform Configuration

```terraform
resource "aws_dynamodb_table_replica" "this" {
  global_table_arn        = var.global_table_arn

  deletion_protection_enabled = true
}
```

### References

https://github.com/hashicorp/terraform-provider-aws/pull/29924

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by andrei-shulaev**
### Description
Adds support to enable/disable deletion protection when creating/updating a dynamodb table replica.

### Relations
Closes #30213 

### Output from Acceptance Testing

```console
make testacc TESTS=TestAccDynamoDBTableReplica_deletion_protection PKG=dynamodb  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccDynamoDBTableReplica_deletion_protection'  -timeout 360m
=== RUN   TestAccDynamoDBTableReplica_deletion_protection
=== PAUSE TestAccDynamoDBTableReplica_deletion_protection
=== CONT  TestAccDynamoDBTableReplica_deletion_protection
--- PASS: TestAccDynamoDBTableReplica_deletion_protection (371.90s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   377.910s
```

```console
make testacc TESTS=TestAccDynamoDBTableReplica_basic PKG=dynamodb                
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccDynamoDBTableReplica_basic'  -timeout 360m
=== RUN   TestAccDynamoDBTableReplica_basic
=== PAUSE TestAccDynamoDBTableReplica_basic
=== CONT  TestAccDynamoDBTableReplica_basic
--- PASS: TestAccDynamoDBTableReplica_basic (203.78s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   209.721s
```

### CODE REVIEW

**Comment by andrei-shulaev** [comment]
Hey @DrFaust92, could you please take a peek?

**Comment by andrei-shulaev** [comment]
@ewbankkit @justinretzolk 
hey could you please have a look at this PR?

**Comment by ankon** [comment]
:heart: This would be great to have, and stop us from manually touching infrastructure!

**Comment by andrei-shulaev** [comment]
@ewbankkit @justinretzolk @DrFaust92

Hey it's been a while without a review, can i get your help on it please?

**Review by aliaksandr-d — APPROVED** [review]

File: internal/service/dynamodb/table_replica.go [hunk 1]
```diff
@@ -417,6 +426,22 @@ func resourceTableReplicaUpdate(ctx context.Context, d *schema.ResourceData, met
...
+			_, err := conn.UpdateTable(ctx, &dynamodb.UpdateTableInput{
+				TableName:                 aws.String(tableName),
+				DeletionProtectionEnabled: aws.Bool(d.Get("deletion_protection_enabled").(bool)),
+			})
+			if err != nil {
+				return create.AppendDiagError(diags, names.DynamoDB, create.ErrActionUpdating, resNameTableReplica, d.Id(), err)
+			}
+			// Deletion protection change is reflected only on table status, and not in replicas status field
+			// There is certain lag after attr update and table status change so doing it with delay
+			if _, err := waitTableActiveAfterDeletionProtectionChange(ctx, conn, tableName, d.Timeout(schema.TimeoutUpdate)); err != nil {
```
Comment (bschaatsbergen) [inline internal/service/dynamodb/table_replica.go:440]: Curious whether we could still make this work with the existing waitTableActive function and an increased timeout, something for one of the core maintainers to check.

File: internal/service/dynamodb/table_replica_test.go [hunk 2]
```diff
@@ -759,3 +795,61 @@ resource "aws_dynamodb_table_replica" "test" {
...
+
+resource "aws_dynamodb_table_replica" "test" {
+  provider                    = "awsalternate"
+  global_table_arn            = aws_dynamodb_table.test.arn
+  deletion_protection_enabled = true
+}
+`, rName))
+}
+
+func testAccTableReplicaConfig_disable_deletion_protection(rName string) string {
```
Comment (bschaatsbergen) [inline internal/service/dynamodb/table_replica_test.go:828]: Would it be possible to combine this and the `testAccTableReplicaConfig_enable_deletion_protection` function and use a boolean to enable/disable deletion protection? It looks like both configurations are very similar and the only difference is setting the `deletion_protection` argument.

**Review by bschaatsbergen — COMMENTED** [review]
Hey @andrei-shulaev, thanks a lot for taking the time to raise this pull request! 👏🏼 

I've left a couple drive-by comments before one of the core maintainers takes a look at this PR. Added, please make sure to update the documentation of this resource and data source with the new argument.

File: internal/service/dynamodb/table_replica.go [hunk 3]
```diff
@@ -90,6 +90,10 @@ func resourceTableReplica() *schema.Resource {
 				ForceNew:         true,
 				ValidateDiagFunc: enum.Validate[awstypes.TableClass](),
 			},
+			"deletion_protection_enabled": { // direct to replica
```
[the remark this answers is not in the document]
Comment (andrei-shulaev) [reply internal/service/dynamodb/table_replica.go]: removed.

File: internal/service/dynamodb/table_replica.go [hunk 4]
```diff
@@ -315,6 +319,10 @@ func resourceTableReplicaReadReplica(ctx context.Context, d *schema.ResourceData
 
 	setTagsOut(ctx, Tags(tags))
 
+	if table.DeletionProtectionEnabled != nil {
```
[the remark this answers is not in the document]
Comment (andrei-shulaev) [reply internal/service/dynamodb/table_replica.go]: yep, added.

File: internal/service/dynamodb/table_replica_test.go (hunk 2 above)
Comment (andrei-shulaev) [reply internal/service/dynamodb/table_replica_test.go]: added

**Comment by bschaatsbergen** [comment]
Hey @andrei-shulaev,

Thank you for your initial work on supporting the `deletion_protection_enabled` argument in the table replica resource. Your analysis on the propagation delay and status reporting was spot on. I’ve made a few minor adjustments, mainly cosmetic changes and tweaks to the call order. Additionally, I added a function that introduces a 1 minute and 30 second delay (just enough for propagation to complete) during the create operation. This improves the performance of subsequent read operations (refreshes), which are more frequent than re-creating the resource. The delay is applied only when the `deletion_protection_enabled` argument is set.

I'm now running the tests, because they take some time I'll provide an update here once they're all ran successfully. 

Thanks again for your work!

**Review by bschaatsbergen — APPROVED** [review]
```
 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_basic' -v
2024/11/28 14:44:32 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_basic
=== PAUSE TestAccDynamoDBTableReplica_basic
=== CONT  TestAccDynamoDBTableReplica_basic
--- PASS: TestAccDynamoDBTableReplica_basic (231.72s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   237.215s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_disappears' -v
2024/11/28 14:53:21 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_disappears
=== PAUSE TestAccDynamoDBTableReplica_disappears
=== CONT  TestAccDynamoDBTableReplica_disappears
--- PASS: TestAccDynamoDBTableReplica_disappears (221.18s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   226.806s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitr' -v
2024/11/28 14:53:26 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitr
=== PAUSE TestAccDynamoDBTableReplica_pitr
=== RUN   TestAccDynamoDBTableReplica_pitrKMS
=== PAUSE TestAccDynamoDBTableReplica_pitrKMS
=== RUN   TestAccDynamoDBTableReplica_pitrDefault
=== PAUSE TestAccDynamoDBTableReplica_pitrDefault
=== CONT  TestAccDynamoDBTableReplica_pitr
=== CONT  TestAccDynamoDBTableReplica_pitrDefault
=== CONT  TestAccDynamoDBTableReplica_pitrKMS
--- PASS: TestAccDynamoDBTableReplica_pitrKMS (255.85s)
--- PASS: TestAccDynamoDBTableReplica_pitr (280.73s)
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (409.53s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   415.134s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitrDefault' -v
2024/11/28 14:52:19 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitrDefault
=== PAUSE TestAccDynamoDBTableReplica_pitrDefault
=== CONT  TestAccDynamoDBTableReplica_pitrDefault
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (351.79s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   357.359s

$ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitrKMS' -v
2024/11/28 14:53:29 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitrKMS
=== PAUSE TestAccDynamoDBTableReplica_pitrKMS
=== CONT  TestAccDynamoDBTableReplica_pitrKMS
--- PASS: TestAccDynamoDBTableReplica_pitrKMS (281.57s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   287.001s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_keys' -v
2024/11/28 15:00:04 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_keys
=== PAUSE TestAccDynamoDBTableReplica_keys
=== CONT  TestAccDynamoDBTableReplica_keys
--- PASS: TestAccDynamoDBTableReplica_keys (405.00s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   410.549s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_tableClass' -v
2024/11/28 15:00:08 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_tableClass
=== PAUSE TestAccDynamoDBTableReplica_tableClass
=== CONT  TestAccDynamoDBTableReplica_tableClass
--- PASS: TestAccDynamoDBTableReplica_tableClass (518.27s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   527.870s

```

**Review by bschaatsbergen — APPROVED** [review]
LGTM! 🚀 (had to run tests separately because of some strange context behavior)

```
 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_basic' -v
2024/11/28 14:44:32 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_basic
=== PAUSE TestAccDynamoDBTableReplica_basic
=== CONT  TestAccDynamoDBTableReplica_basic
--- PASS: TestAccDynamoDBTableReplica_basic (231.72s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   237.215s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_disappears' -v
2024/11/28 14:53:21 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_disappears
=== PAUSE TestAccDynamoDBTableReplica_disappears
=== CONT  TestAccDynamoDBTableReplica_disappears
--- PASS: TestAccDynamoDBTableReplica_disappears (221.18s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   226.806s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitrDefault' -v
2024/11/28 14:52:19 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitrDefault
=== PAUSE TestAccDynamoDBTableReplica_pitrDefault
=== CONT  TestAccDynamoDBTableReplica_pitrDefault
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (351.79s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   357.359s

$ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitrKMS' -v
2024/11/28 14:53:29 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitrKMS
=== PAUSE TestAccDynamoDBTableReplica_pitrKMS
=== CONT  TestAccDynamoDBTableReplica_pitrKMS
--- PASS: TestAccDynamoDBTableReplica_pitrKMS (281.57s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   287.001s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_pitrDefault' -v
2024/11/28 14:52:19 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_pitrDefault
=== PAUSE TestAccDynamoDBTableReplica_pitrDefault
=== CONT  TestAccDynamoDBTableReplica_pitrDefault
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (351.79s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   357.359s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_tableClass' -v
2024/11/28 15:00:08 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_tableClass
=== PAUSE TestAccDynamoDBTableReplica_tableClass
=== CONT  TestAccDynamoDBTableReplica_tableClass
--- PASS: TestAccDynamoDBTableReplica_tableClass (518.27s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   527.870s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_keys' -v
2024/11/28 15:00:04 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_keys
=== PAUSE TestAccDynamoDBTableReplica_keys
=== CONT  TestAccDynamoDBTableReplica_keys
--- PASS: TestAccDynamoDBTableReplica_keys (405.00s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   410.549s

 $ go test ./internal/service/dynamodb/... -v -count 1 -parallel 4 -run='TestAccDynamoDBTableReplica_deletionProtection' -v
2024/11/28 15:00:12 Initializing Terraform AWS Provider...
=== RUN   TestAccDynamoDBTableReplica_deletionProtection
=== PAUSE TestAccDynamoDBTableReplica_deletionProtection
=== RUN   TestAccDynamoDBTableReplica_deletionProtectionDefault
=== PAUSE TestAccDynamoDBTableReplica_deletionProtectionDefault
=== CONT  TestAccDynamoDBTableReplica_deletionProtection
=== CONT  TestAccDynamoDBTableReplica_deletionProtectionDefault
--- PASS: TestAccDynamoDBTableReplica_deletionProtectionDefault (512.04s)
--- PASS: TestAccDynamoDBTableReplica_deletionProtection (512.90s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   525.348s
```

Since I’ve made some updates to the code in this PR, I’m requesting another review from the core team. Thanks for your patience!

**Review by jar-b — APPROVED** [review]
LGTM 🎉 

```console
% make testacc PKG=dynamodb TESTS=TestAccDynamoDBTableReplica_
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.3 test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccDynamoDBTableReplica_'  -timeout 360m
2024/12/03 11:14:52 Initializing Terraform AWS Provider...

--- PASS: TestAccDynamoDBTableReplica_pitr (260.06s)
=== CONT  TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_DefaultTag
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_emptyResourceTag (266.37s)
=== CONT  TestAccDynamoDBTableReplica_basic
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnUpdate_Replace (267.96s)
=== CONT  TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_ResourceTag
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_updateToProviderOnly (270.02s)
=== CONT  TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Add
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nullOverlappingResourceTag (272.81s)
=== CONT  TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Replace
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nullNonOverlappingResourceTag (273.78s)
=== CONT  TestAccDynamoDBTableReplica_tags_EmptyMap
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_emptyProviderOnlyTag (275.60s)
=== CONT  TestAccDynamoDBTableReplica_tags_AddOnUpdate
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_updateToResourceOnly (276.02s)
=== CONT  TestAccDynamoDBTableReplica_tags_null
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnCreate (277.72s)
=== CONT  TestAccDynamoDBTableReplica_tags_ComputedTag_OnCreate
--- PASS: TestAccDynamoDBTableReplica_disappears (290.68s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_overlapping (378.84s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_providerOnly (380.90s)
--- PASS: TestAccDynamoDBTableReplica_pitrKMS (383.09s)
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (383.75s)
--- PASS: TestAccDynamoDBTableReplica_keys (394.85s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nonOverlapping (402.72s)
--- PASS: TestAccDynamoDBTableReplica_tags (404.44s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnUpdate_Add (412.77s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyMap (235.27s)
--- PASS: TestAccDynamoDBTableReplica_tags_null (234.32s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Add (242.74s)
--- PASS: TestAccDynamoDBTableReplica_tags_AddOnUpdate (237.38s)
--- PASS: TestAccDynamoDBTableReplica_tableClass (517.17s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnCreate (240.49s)
--- PASS: TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_DefaultTag (258.90s)
--- PASS: TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_ResourceTag (255.81s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Replace (253.21s)
--- PASS: TestAccDynamoDBTableReplica_basic (260.55s)
--- PASS: TestAccDynamoDBTableReplica_deletionProtection (760.28s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   766.869s
```

**Comment by jar-b** [comment]
Thanks for your contribution, @andrei-shulaev! 👍

**Review by gdavison — APPROVED** [review]
Looks good! 🚀 

--- PASS: TestAccDynamoDBTableReplica_disappears (211.09s)
--- PASS: TestAccDynamoDBTableReplica_pitr (246.23s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Replace (299.96s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nullOverlappingResourceTag (301.63s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_emptyProviderOnlyTag (309.08s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_updateToProviderOnly (317.87s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_updateToResourceOnly (321.74s)
--- PASS: TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_DefaultTag (322.95s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnUpdate_Add (323.12s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_emptyResourceTag (324.06s)
--- PASS: TestAccDynamoDBTableReplica_basic (326.05s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nullNonOverlappingResourceTag (332.96s)
--- PASS: TestAccDynamoDBTableReplica_pitrKMS (355.25s)
--- PASS: TestAccDynamoDBTableReplica_pitrDefault (381.77s)
--- PASS: TestAccDynamoDBTableReplica_keys (426.55s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_overlapping (429.84s)
--- PASS: TestAccDynamoDBTableReplica_tags_IgnoreTags_Overlap_ResourceTag (431.03s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_nonOverlapping (432.09s)
--- PASS: TestAccDynamoDBTableReplica_tags (435.69s)
--- PASS: TestAccDynamoDBTableReplica_tags_ComputedTag_OnCreate (236.15s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnUpdate_Add (251.82s)
--- PASS: TestAccDynamoDBTableReplica_tags_null (230.41s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnCreate (345.70s)
--- PASS: TestAccDynamoDBTableReplica_tags_DefaultTags_providerOnly (313.03s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyMap (240.15s)
--- PASS: TestAccDynamoDBTableReplica_tags_AddOnUpdate (243.73s)
--- PASS: TestAccDynamoDBTableReplica_tags_EmptyTag_OnUpdate_Replace (271.74s)
--- PASS: TestAccDynamoDBTableReplica_deletionProtection (781.69s)
--- PASS: TestAccDynamoDBTableReplica_tableClass (475.17s)

**Pull request merged**